### PR TITLE
Support true 16-bit precision with deepspeed in Trainer

### LINKED ISSUE #17727: Cast fp tensor to half when using 16-true precision
### Description & Motivation

The goal of this feature is to support transferring tensors in half precision to an accelerator when using 16-true precision.
Currently, some strategies support this behavior but only for 16-mixed of bf16-mixed, such as DeepSpeed and IPUStrategy.
This is done in `batch_to_device` method which in turn calls `_fp_to_half`.
The definition of done for this PR is to extend the `_fp_to_half` method.

### Pitch

By changing the behavior of `_fp_to_half` when using 16-true precision, tensors are cast to half before being transferred to the accelerator(s).

Should be a matter of changing the method to:
```
def _fp_to_half(
    tensor: Tensor,
    precision: Literal[
        "64-true",
        "32-true",
        "16-mixed",
        "bf16-mixed",
        "16-true",
    ],
) -> Tensor:
    if str(precision) in ("16-mixed", "16-true"):
        return _convert_fp_tensor(tensor, torch.half)
    if precision == "bf16-mixed":
        return _convert_fp_tensor(tensor, torch.bfloat16)
    return tensor
```

As far as I understand the _fp_to_half method is only used inside  `batch_to_device` so I don't expect any side-effect from this change. Hopefully that is really the case.

### Alternatives

I believe the alternative is to cast tensors to half precision inside the datamodule, which involves making the datamodule aware of the training precision, which might not be trivial to do.
Another alternative is to cast tensors after they are in the accelerator. This involves changing the model to pre-append a casting operation. This seems easier, but isn't exactly the same as the proposal above. By transferring data in half precision, the data payload is reduced which increases the performance of the training loop. 

### Additional context

_No response_

cc @borda

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Follow up to #17609 

Fixes #17727
Fixes #17728 

Requires https://github.com/Lightning-AI/lightning-Graphcore/pull/32
Similar changes as in #17576

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @awaelchli

### CODE REVIEW

File: src/lightning/pytorch/strategies/deepspeed.py [hunk 1]
```diff
@@ -893,6 +890,9 @@ def register_strategies(cls, strategy_registry: _StrategyRegistry) -> None:
             offload_optimizer_device="nvme",
         )
 
+    # TODO: discuss should we keep overriding this?
     def batch_to_device(self, batch: Any, device: Optional[torch.device] = None, dataloader_idx: int = 0) -> Any:
```
Comment (awaelchli) [inline src/lightning/pytorch/strategies/deepspeed.py:894]: Arguments for removing the override: 
- Consistent with other strategies who convert the dtype in the `*_step()` methods
- Otherwise we would be calling `precision_plugin.convert_input(batch)` twice redundantly
- Even if the user overrides the `transfer_batch_to_device` hook and changes the dtype, the dtype would be changed back to what the precision plugin wants before passing it into the LM-step methods.
- The previous behavior could be considered a bug, because the docs of the `transfer_batch_to_device` say:

       

	>  The returned batch is of the same type as the input batch, just
	>         having all tensors on the correct device.
	and
	
	> This hook should only transfer the data and not modify it


Arguments for keeping it:
- Could be considered a breaking change for deepspeed users who expected the dtype to be already converted in `transfer_batch_to_device`

**Review by justusschock — APPROVED** [review]

File: requirements/_integrations/accelerators.txt [hunk 2]
```diff
@@ -1,3 +1,3 @@
 # validation HPU connectors
 lightning-habana >=0.1.0
-lightning-graphcore >=0.1.0.rc3
+git+https://github.com/Lightning-AI/lightning-Graphcore.git@remove-protected-import  # TODO: revert
```
Comment (Borda) [inline requirements/_integrations/accelerators.txt:3]:
```suggestion
lightning-graphcore >=0.1.0.rc4
```

**Review by Borda — APPROVED** [review]

File: src/lightning/pytorch/strategies/deepspeed.py (hunk 1 above)
Comment (Borda) [reply src/lightning/pytorch/strategies/deepspeed.py:894]: Will this be follow-up work?

Comment (awaelchli) [reply src/lightning/pytorch/strategies/deepspeed.py:894]: I think we should make a decision here. Waiting for @carmocca's thoughts :)

**Review by carmocca — APPROVED** [review]

File: src/lightning/pytorch/strategies/deepspeed.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning/pytorch/strategies/deepspeed.py:894]: Yes, the batch gets first moved to the device, then passed to the Strategy.training_step methods where the dtype conversion happens (if necessary).

Comment (awaelchli) [reply src/lightning/pytorch/strategies/deepspeed.py:894]: We could also do it the other way around, first converting to dtype, then moving to device. This would then be a change that affects all other strategies. Note that in Fabric we do the casting also after device transfer, because the user is generally responsible for making sure the data is on the device before passing it into the model.

Comment (awaelchli) [reply src/lightning/pytorch/strategies/deepspeed.py]: I'm keeping it for now, not to regress on memory and since there is an explicit test case (thanks for pointing to the PR). I'm following up with a PR that standardizes this.
